static/ts/dict.ts

```diff
@@ -0,0 +1,154 @@
+        return dict;
+    }
+
+    /**
+     * Construct a Dick object from an array with each element set to `true`.
+     * @param arr - An array of keys
+     * @param opts - Options to be passed to the Dict constructor
+     */
+    static from_array(arr: any[], opts?: object): Dict {
+        if (!(arr instanceof Array)) {
+            throw new TypeError("Argument is not an array");
+        }
+
+        let dict = new Dict(opts);
+        for (const key of arr) {
+            dict.set(key, true);
+        }
+        return dict;
+    }
+
+    private _munge(key: any): string | undefined {
+        if (key === undefined) {
+            blueslip.error("Tried to call a Dict method with an undefined key.");
+            return undefined;
+        }
+        key = ':' + key.toString();
+        return this._opts.fold_case ? key.toLowerCase() : key;
+    }
+
+    clone(): Dict {
+        let dict = new Dict(this._opts);
+        dict._items = { ...this._items };
+        return dict;
+    }
+
+    get(key: any): any {
+        const mapping = this._items[this._munge(key)];
+        if (mapping === undefined) {
+            return undefined;
+        }
+        return mapping.v;
+    }
+
+    set(key: any, value: any): any {
+        this._items[this._munge(key)] = {k: key, v: value};
+        return value;
+    }
+
+    /**
+     * If `key` exists in the Dict, return its value. Otherwise insert `key`
+     * with a value of `value` and return the value.
+     */
+    setdefault(key: any, value: any): any {
+        const mapping = this._items[this._munge(key)];
+        if (mapping === undefined) {
+            return this.set(key, value);
+        }
+        return mapping.v;
+    }
+
+    has(key: any): boolean {
+        return this._items.hasOwnProperty(this._munge(key));
+    }
+
+    del(key: any): boolean {
+        // `delete` could throw a TypeError in strict mode.
+        try {
+            delete this._items[this._munge(key)];
+        } catch (_e) {
+            return false;
+        }
+        return true;
+    }
+
+    keys(): string[] {
+        return _.pluck(_.values(this._items), 'k');
+    }
+
+    values(): any[] {
+        return _.pluck(_.values(this._items), 'v');
+    }
+
+    items(): [string, any][] {
+        return _.values(this._items).map(
+            (mapping: KV): [string, any] => [mapping.k, mapping.v]
+        );
+    }
+
+    num_items(): number {
+        return Object.keys(this._items).length;
+    }
+
+    is_empty(): boolean {
+        return _.isEmpty(this._items);
+    }
+
+    each(f: (v: any, k?: string) => void) {
+        _.each(this._items, (mapping: KV) => f(mapping.v, mapping.k));
+    }
+
+    clear() {
+        this._items = {};
+    }
+}
+
+// Using `export =` here instead of default export to avoid writing
+// `require(...).default` when importing from a CommonJS module.
+// TODO: switch to default export when the whole js codebase have been migrated
+// to Typescript.
+export = Dict;
```

Comment:
I don’t see a reason to do multiple migrations here—nothing wrong with writing `require(…).default`. Although actually, I’d actually prefer to avoid default exports entirely, and write `export class Dict {…}` and `require(…).Dict`.